event-web-framework/draft.markdown

```diff
@@ -0,0 +1,872 @@
+
+In other words, the above is equivalent to
+
+```lisp
+	(let ((it (cdr (assoc :room parameters))))
+	  (if it
+	      (uri-decode it)
+	      (error (make-instance 'http-assertion-error :assertion 'room))))
+```
+
+Strongly-typed functional languages like Haskell often use a `Maybe` type in this situation. In Common Lisp, we capture the symbol `it` in the expansion as the name for the result of the check. 
+
+Understanding this, we should be able to see that `arg-exp` is generating a specific, repetitive, piece of the code tree that we eventually want to evaluate. In this case, the piece that checks for the presence of the given parameter among the handlers' `parameters`. Now, let's move onto...
+
+#### type-expression
+
+```lisp
+	(defgeneric type-expression (parameter type)
+	  (:documentation
+	   "A type-expression will tell the server how to convert a parameter from a string to a particular, necessary type."))
+    ...
+	(defmethod type-expression (parameter type) nil)
+```
+
+This is a generic function that generates new tree structures (coincidentally Lisp code), rather than just a function. The only thing the above tells you is that by default, a `type-expression` is `NIL`. Which is to say, we don't have one. If we encounter a `NIL`, we use the raw output of `arg-exp`, but that doesn't tell us much about the most common case. To see that, lets take a look at a built-in (to `:house`) `define-http-type` expression.
+
+```lisp
+	(define-http-type (:integer)
+	    :type-expression `(parse-integer ,parameter :junk-allowed t)
+		:type-assertion `(numberp ,parameter))
+```
+
+An `:integer` is something we're making from a `parameter` by using `(parse-integer parameter :junk-allowed t)`. `junk-allowed` tells `parse-integer` that we're not confident the data we're giving it is actually parseable, so we need to make sure that the returned result is an integer. If it isn't, we get this behaviour: 
+
+```
+	HOUSE> (type-expression 'blah :integer)
+	(PARSE-INTEGER BLAH :JUNK-ALLOWED T)
+	HOUSE> 
+```
+
+`define-http-handler`[^readable] is one of the exported symbols for our framework. This lets our application programmers define their own types to simplify parsing above the handful of "builtins" that we give them (`:string`, `:integer`, `:keyword`, `:json`, `:list-of-keyword` and `:list-of-integer`.) 
+
+```lisp
+	(defmacro define-http-type ((type) &key type-expression type-assertion)
+	  (with-gensyms (tp)
+	    `(let ((,tp ,type))
+	       ,@(when type-expression
+		       `((defmethod type-expression (parameter (type (eql ,tp)))
+			   ,type-expression)))
+	       ,@(when type-assertion
+		       `((defmethod type-assertion (parameter (type (eql ,tp)))
+			   ,type-assertion))))))
+```
+
+[^readable]: This macro is difficult to read because it tries hard to make its output human-readable, by expanding `NIL`s away using `,@` where possible.
```

Comment:
May want to expand on this by showing an expansion or two, given that this is already a footnote.